lib/ukconsole: Add generic console device interface

thass0 opened the pull request:
This PR introduces `ukconsole`, a library that exposes a generic interface for console devices. This is groundwork for adding the gdb stub to Unikraft.

### Prerequisite checklist

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.uk`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.uk) on your commit series before opening this PR;
 - [x] Updated relevant documentation.

Changes are tested on KVM using QEMU (both x86_64 and ARM64) and on Xen (x86_64).

### Base target

 - Architecture(s): N/A
 - Platform(s): `kvm`, `xen`
 - Application(s): N/A

### Additional configuration

The `app-helloworld` and `catalog/library/nginx/1.25` projects should both build without changing the configuration

### Description of changes

From the perspective of user libraries, all "print" and "read input" operations for consoles now go through the `ukconsole` interface (`console.h`). Before, the `ukplat_[cout|cin][k|d]` functions were used for that purpose. These `ukplat` functions differentiated between kernel and debug printing. This difference is gone with `ukconsole`. kernel/debug printing wasn't really used so the difference will be phased out over time.

Drivers can register themselves with `ukconsole` in one of two ways: statically or dynamically. All statically registered device drivers are initialized at once with the `uk_console_early_init` function. Dynamic registration can happen at any point in time. `ukconsole` expects a dynamically registered device to be initialized before it's registered. The distinction between dynamically and statically registered devices is useful because some device drivers (e.g., PL011, NS16550) need information that becomes available later during boot before they can be initialized. Other drivers (e.g., COM) have no such prerequisites and can be initialized very early on.

The PL011 and NS16550 drivers are adapted to use `ukconsole`. The COM driver and the VGA driver from `plat/kvm/` were moved to `drivers/uktty` and also now use `ukconsole`. The Xen console drivers were moved to `drivers/xen` and updated to use `ukconsole`.

Note that some changes introduced early on are removed in later commits. Namely, the dynamic registration mechanism is simplified a lot later on. The string console device names used at the start are also removed entirely later. 

GitHub-Depends-On: #1473

thass0 committed:
lib/posix-tty: Replace `ukplat_[coutk|cink]` with `ukconsole` calls

This commit makes `posix-tty` call into `ukconsole` instead of the
platform-provided print system.

Signed-off-by: Thassilo Schulze <thassilo@unikraft.io>

thass0 committed:
include/plat: Delete old `ukplat_[cout|cin]` function declarations

There functions are now replaced by the `ukconsole` interface
and the drivers that implemented them are moved to `drivers/uktty`.

Signed-off-by: Thassilo Schulze <thassilo@unikraft.io>

thass0 committed:
plat/kvm: Move VGA driver from plat/ to drivers/

This commit moves the VGA driver to `drivers/ukconsole`. In future
commits, the VGA driver will be adapted to work with `ukconsole`.

Signed-off-by: Thassilo Schulze <thassilo@unikraft.io>

thass0 committed:
plat/kvm: Delete console driver files for x86_64

* The implementation of the console drivers now lives
  in `drivers/`. Additional functionality that was in
  the platform before is in `ukconsole`.
* The delete console driver configuration options now
  live in the `drivers/uktty/` subdirectories
* Delete MRD-related code for x86 video memory. This
  MRD is inserted by the VGA driver on init now.

Signed-off-by: Thassilo Schulze <thassilo@unikraft.io>

thass0 committed:
lib/ukdebug: Switch from `ukplat` to `ukconsole` for printing

For printing debug messages, `ukdebug` now uses `ukconsole`
instead of the old printing functions from the platform.

Checkpatch-Ignore: SPACING
Checkpatch-Ignore: LONG_LINE
Signed-off-by: Thassilo Schulze <thassilo@unikraft.io>

thass0 committed:
lib/syscall_shim: Switch from `ukplat` to `ukconsole` for printing

This commit makes the syscall shim call into `ukconsole`
to print some message. Direct calls to `ukconsole` like
this one should normally be avoided in favor of calling
into `ukdebug`. But in this case, we don't want to print
the extra information that `ukdebug` always includes.

When there is a kernel print system in the future, the
new print calls in this commit will be updated.

This commit avoids creating a direct dependency between
the syscall shim and `ukconsole` because the call into
`ukconsole` is a workaround anyway and will be removed
in the future.

Signed-off-by: Thassilo Schulze <thassilo@unikraft.io>

thass0 committed:
lib/devfs: Switch from `ukplat` to `ukconsole` for printing

`ukconsole` replaces the use of the platform-provided print
functions so `devfs` should use `ukconsole` too.

Signed-off-by: Thassilo Schulze <thassilo@unikraft.io>

thass0 committed:
plat/xen: Move console driver code into drivers/

Just like the drivers from the KVM platform, we also
want to move the Xen console drivers to `drivers/`
and later register them with `ukconsole`.

Signed-off-by: Thassilo Schulze <thassilo@unikraft.io>

thass0 committed:
plat/xen: Update Xen to use console drivers in `drivers/`

* Imply the console drivers from the Xen platform
  if `ukconsole` is enabled.
* The original Xen console header that was called
  `common/console.h` in the platform now lives in
  the `drivers/` folder where it's accessible as
  `xen/console.h`.
* The Xen console drivers in the `drivers/` directory
  will register in the early init tab. That's why the
  setup on x86 and Arm has to change slightly (see diff)

Signed-off-by: Thassilo Schulze <thassilo@unikraft.io>

thass0 committed:
drivers/xen/console: Adapt HV console to work with `ukconsole`

The standard HV console now works with `ukconsole`. It registers
its initialization function in the early init table.

Checkpatch-Ignore: CAMELCASE
Signed-off-by: Thassilo Schulze <thassilo@unikraft.io>

thass0 committed:
lib/nolibc: Switch from `ukplat` to `ukconsole` for printing

`ukconsole` replaces the use of the platform-provided print
functions so `nolibc` should use `ukconsole`.

Signed-off-by: Thassilo Schulze <thassilo@unikraft.io>

thass0 committed:
plat/kvm: Move `serial_console.c` into `drivers/ukconsole/ns16550/`

`serial_console.c` implements a driver for the COM ports
on x86. The driver is now moved out of the platform and
will be registered with `ukconsole` in the future. The device
underlying the COM port interface is a 16550 / 8250 so the
`ns16550` subdirectory is the right place to put this driver.
Because the device is the same on Arm and on x86, the `com.c`
and `ns16550.c` driver should be merged into one driver in
the future.

Signed-off-by: Thassilo Schulze <thassilo@unikraft.io>

thass0 committed:
lib/ukconsole: Add abstraction layer for console drivers

`ukconsole` is a new library that implements a generic
interface for console devices. It will be used to replace
the current kernel/debug print system that's based on the
`ukplat_c[out|in][k|d]` functions. `ukconsole` is
groundwork for properly integrating the gdb stub with
Unikraft and for adding a proper kernel print system
later on.

Signed-off-by: Thassilo Schulze <thassilo@unikraft.io>

thass0 committed:
drivers/ukconsole: Adapt PL011 driver to work with `ukconsole`

The PL011 driver can register devices in the early init table
and in the inittab. There can only be a single device that's
registers in the early init table and the base address for
that device can be configured using kconfig or using a kernel
parameter.

Any number of devices can be registered in the init table. The
DTB is searched for PL011 UARTs. If a device is found, a new
device structure is dynamically allocated and registered with
`ukconsole`.

Signed-off-by: Thassilo Schulze <thassilo@unikraft.io>

thass0 committed:
drivers/ukconsole: Adapt NS16550 for use with `ukconsole`

These changes are functionally identical to those in the
PL011 driver.

Signed-off-by: Thassilo Schulze <thassilo@unikraft.io>

thass0 committed:
drivers/ukconsole: Adapt COM driver to work with `ukconsole`

The COM driver now exposes COM ports COM1 to COM4. The
driver registers a `struct uk_console` for a port to
make that port available to `ukconsole`. One of the
ports can be configured as the "early console". This
port is initialized and registered in the early init
table. All other ports besides the configured "early
console" are initialized and registered when the init
table runs.

Checkpatch-Ignore: EMBEDDED_FUNCTION_NAME
Signed-off-by: Thassilo Schulze <thassilo@unikraft.io>

thass0 committed:
plat/kvm: Imply COM drivers in x86 platform configuration

The KVM platform configuration for x86 now implies the
COM driver in `drivers/`.

Signed-off-by: Thassilo Schulze <thassilo@unikraft.io>

thass0 committed:
drivers/ukconsole: Adapt VGA driver for use with `ukconsole`

The VGA driver registers a device with `ukconsole` in
the early init table. The device that the driver registers
only has the `UK_CONSOLE_FLAG_STDOUT` flag set because the
VGA driver doesn't implement keyboard input.

Signed-off-by: Thassilo Schulze <thassilo@unikraft.io>

michpappas commented on plat/common/include/uk/plat/common/common.lds.h:
```diff
@@ -122,12 +122,12 @@
 		KEEP(*(SORT_BY_NAME(.uk_inittab[1-6][0-9])))		\
-	}								\
-	uk_inittab_end = .;
+		uk_inittab_end = .;					\
```
I would move this change to a separate PR as it's not immediately related to the scope of the current one (which is big enough already). Saying that, you should probably apply the same change to all sections in `common.ld.h`. Moreover, you could also moved `.uk_inittab` to a library-defined section in `libukboot`.

michpappas commented on lib/nolibc/scanf.c:
```diff
@@ -66,6 +66,9 @@ vfscanf(FILE *fp, const char *fmt, va_list ap)
 	if (ret < 0)
 		return ret;
 
+	size = size < sizeof(buf) ? size : sizeof(buf) - 1;
```
I would open a separate PR for this one too.

thass0 replied:
I opened a separate PR here: https://github.com/unikraft/unikraft/pull/1474

michpappas commented on lib/ukconsole/Config.uk:
```diff
@@ -0,0 +1,10 @@
+menuconfig LIBUKCONSOLE
+	bool "ukconsole: generic console driver interface"
+	default n
```
`default n` is redundant as `n` is the default value for bool symbols in Kconfig: https://docs.zephyrproject.org/latest/build/kconfig/tips.html#redundant-defaults

The same comment applies to `LIBUKCONSOLE_TEST` below.

michpappas commented on lib/ukconsole/Makefile.uk:
```diff
@@ -0,0 +1,10 @@
+CINCLUDES-$(CONFIG_LIBUKCONSOLE) += -I$(LIBUKCONSOLE_BASE)/include
+CXXINCLUDES-$(CONFIG_LIBUKCONSOLE) += -I$(LIBUKCONSOLE_BASE)/include
+
+LIBUKCONSOLE_SRCS-y += $(LIBUKCONSOLE_BASE)/console.c|isr
```
Print functions are generally not async-safe. Is there a reason for setting the `isr` flag here?
The same question applies to tests below.

thass0 replied:
The console functions may run very early in the boot sequence. At that point , we don't want to assume that the FPU, SIMD, etc. are properly initialized. Compiling ISR safe keeps to compiler from using SIMD instructions for example. You, @michpappas , said before that in practice, these parts of the CPU are already initialize when the console functions run. Maybe compiling them ISR safe is premature future-proofing. @skuenzer What do you think

michpappas replied:
Indeed. Right now everything is initialized before early init. Since initializing SIMD in `lcpu_init()` is something TBD I would save associated changes to that time. @skuenzer is that ok with you?

skuenzer replied:
This is a very good point. So far, we used the `|isr` approach for early boot code but you are right @michpappas, actually not needed. I think it would make sense if we deploy the following convention/assumption: Any CPU feature requested by compiler (via selected target CPU via `arch`) is initialized in early assembly code, if missing or fails, we stop boot with an very short and static error message, like `"Unsupported CPU"`. Any additional feature, we should initialize later during boot and provide an option if this should happen: Use case is `elfloader` where we run binary code that could be compiled differently than the native Unikraft kernel.
I think we should aim this structure with the new `lib/ukboot` concept, and then `|isr` is not needed in fact. Problem is that we have this not yet documented.
Because of this, I am okay with two options, for now:
- We document (probably) in the assembler code that we have to initialize compiler-requested CPU features before jumping into C code; and we remove `|isr` here.
- We keep `|isr` and document here a `TODO/FIXME` that says that `|isr` can be removed as soon we have the just mentioned scheme officially in place.

michpappas replied:
That's fine, in that case we'll drop the `isr` flag and go with a comment. I will update #1472 that introduces the earlytab as that seems a more appropriate place.

michpappas commented on lib/ukconsole/include/uk/console.h:
```diff
@@ -0,0 +1,135 @@
+ * You may not use this file except in compliance with the License.
+ */
+
+/* `ukconsole` functions for user libraries, i.e., libraries that want to
```
Since this is the main header by definition it provides the main library API, so this comment is probably redundant.

michpappas commented on lib/ukconsole/include/uk/console.h:
```diff
@@ -0,0 +1,135 @@
+extern "C" {
+#endif
+
+typedef __ssz (*uk_console_out_t)(const char *buf, __sz len);
```
According to https://unikraft.org/docs/contributing/coding-conventions#typedefs, the convention is to append function definitions with `_func` instead of `_t`.

michpappas commented on lib/ukconsole/include/uk/console.h:
```diff
@@ -0,0 +1,135 @@
+	uk_console_in_t in;
+};
+
+#define UK_CONSOLE_FLAG_STDOUT UK_BIT(0)
```
Preprocessor definition values should be aligned with tabs:
```
#define UK_CONSOLE_FLAG_STDOUT        UK_BIT(0)
#define UK_CONSOLE_FLAG_STDIN         UK_BIT(1)
```
Also given that you only have two values that are 0 and 1 using `UK_BIT()` seems a bit excessive.

thass0 replied:
I think it's fair to stick with `UK_BIT` to be explicit about using bit flags.

michpappas commented on lib/ukconsole/include/uk/console.h:
```diff
@@ -0,0 +1,135 @@
+	UK_TAILQ_ENTRY(struct uk_console) _list;
+};
+
+#define UK_CONSOLE_OPS(_out, _in) \
```
Backslashes in macros should be aligned with tabs:
```
#define UK_CONSOLE_OPS(_out, _in)                              \
        (struct uk_console_ops) { .out = (_out), .in = (_in) }

#define UK_CONSOLE(_name, _ops, _flags)                        \
       (struct uk_console) {                                   \
               .id = 0,                                        \
               .name = (_name),                                \
               .flags = _flags,                                \
               .ops = (_ops),                                  \
               ._list = {0}                                    \
        }
```

michpappas commented on lib/ukconsole/include/uk/console.h:
```diff
@@ -0,0 +1,135 @@
+
+/**
+ * Same as uk_console_out except that dev is used as the console
+ * device to write to instead of using the default device.
```
I would still fully document this and `uk_console_out_with()` as doxygen will autogenerate documentation in the future.

thass0 replied:
Updated the docs comments for `uk_console_in` and `uk_console_out` also.

michpappas commented on lib/ukconsole/include/uk/console.h:
```diff
@@ -0,0 +1,135 @@
+}
+
+/**
+ * @return Number of devices registered with `ukconsole`
```
Summary line missing here.

michpappas replied:
Just to clarify: doxygen provides a _brief_ description and a _long_ description. These are formatted as:
```
/**
 * Brief description (one line)
 * <blank line>
 * Long description (multiple lines)
 * <blank line>
 * Parameters, return value
 * /
```
Although both optional, defining both generates better documentation. In the cases where a function is really simple, we omit the long description.

thass0 replied:
Updated this

michpappas commented on lib/ukconsole/include/uk/console.h:
```diff
@@ -0,0 +1,135 @@
+__u16 uk_console_get_device_count(void);
+
+/**
+ * Register a console device driver. The device driver must be initialized
```
This one also needs a summary line and a description of the parameter.

michpappas commented on lib/ukconsole/console.c:
```diff
@@ -0,0 +1,94 @@
+			return;
+	}
+
+	if (dev->flags) {
```
This is a checkpatch error:
```
WARNING:BRACES: braces {} are not necessary for single statement blocks
```

michpappas commented on lib/ukconsole/include/uk/console.h:
```diff
@@ -0,0 +1,135 @@
+	struct uk_console_ops *ops;
+	const char *name; /* Optional */
+	int flags;
+	UK_TAILQ_ENTRY(struct uk_console) _list;
```
This causes a checkpatch false-positive:
```
CHECK:SPACING: No space is necessary after a cast
#232: FILE: lib/ukconsole/include/uk/console.h:39:
+	UK_TAILQ_ENTRY(struct uk_console) _list;
```

 You should add a line in your commit message to suppress it:
```
Checkpatch-Ignore: SPACING
```

michpappas commented on lib/ukconsole/include/uk/console.h:
```diff
@@ -0,0 +1,135 @@
+ */
+void uk_console_register(struct uk_console *dev);
+
+
```
As checkpatch suggests:
```
CHECK:LINE_SPACING: Please don't use multiple blank lines
#323: FILE: lib/ukconsole/include/uk/console.h:130:
```

michpappas commented on lib/ukconsole/include/uk/console.h:
```diff
@@ -0,0 +1,135 @@
+	__u16 id;
+	struct uk_console_ops *ops;
+	const char *name; /* Optional */
+	int flags;
```
This causes a checkpatch false positive:
```
WARNING:LINE_SPACING: Missing a blank line after declarations
#232: FILE: lib/ukconsole/include/uk/console.h:39:
+	int flags;
+	UK_TAILQ_ENTRY(struct uk_console) _list;
```
You should add a line in your commit message to suppress it:
```
Checkpatch-Ignore: SPACING
```

michpappas commented on lib/ukconsole/console.c:
```diff
@@ -0,0 +1,94 @@
+#include <uk/console.h>
+
+/* List of dynamically registered devices */
+UK_TAILQ_HEAD(uk_console_list, struct uk_console);
```
Here and elsewhere: Unless there is a reason you should probably use `UK_LIST` instead of the old `UK_TAILQ` API. Also notice that this file includes `uk/list.h` while `UK_TAILQ` is defined in `uk/compat_list.h`. The only reason you don't get a compile error is the following line in `list.h`:
```
/* TODO: get rid of the old linked list implementation */
#include <uk/compat_list.h>
```

michpappas commented on lib/ukconsole/console.c:
```diff
@@ -0,0 +1,94 @@
+		uk_console_set_standard_once = 1;
+	}
+
+	if (!uk_console_set_standard_once && !dev->flags) {
```
It would be worth to add a comment to explain the logic both on this block and the one above.

michpappas commented on lib/ukconsole/console.c:
```diff
@@ -0,0 +1,94 @@
+	struct uk_console *dev = __NULL;
+
+	UK_TAILQ_FOREACH(dev, &uk_console_dev_list, _list) {
+		if (dev->flags & UK_CONSOLE_FLAG_STDOUT && dev->ops->out)
```
For readability it's a good practice to group operations with parenthesis regardless of the operator precedence. Similarly for `uk_console_in()`.

michpappas commented on lib/nolibc/scanf.c:
```diff
@@ -4,24 +4,29 @@
 #include <stdarg.h>
 
-#include <uk/plat/console.h>
+#ifdef CONFIG_LIBUKCONSOLE
```
Checks against Kconfig options should use `#if` instead of `#ifdef`.

michpappas replied:
This one is still `#ifdef`

thass0 replied:
Oops missed that

michpappas commented on lib/nolibc/stdio.c:
```diff
@@ -434,8 +436,10 @@ int sprintf(char *str, const char *fmt, ...)
+int vfprintf(FILE *fp __maybe_unused, const char *fmt __maybe_unused,
+	     va_list ap __maybe_unused)
 {
+#ifdef CONFIG_LIBUKCONSOLE
```
`#if`

michpappas commented on lib/nolibc/stdio.c:
```diff
@@ -513,22 +518,23 @@ int putchar(int c)
+	       FILE *restrict stream __maybe_unused,
+	       int newline __maybe_unused)
 {
+#ifdef CONFIG_LIBUKCONSOLE
```
`#if`

michpappas commented on lib/nolibc/scanf.c:
```diff
@@ -4,24 +4,29 @@
-uk_scanf(void *buffer,  size_t *cnt)
+uk_scanf(void *buffer __maybe_unused, size_t *cnt)
 {
+#ifdef CONFIG_LIBUKCONSOLE
```
`#if`

michpappas commented on lib/nolibc/scanf.c:
```diff
@@ -49,6 +54,10 @@ uk_scanf(void *buffer,  size_t *cnt)
+#else /* !CONFIG_LIBUKCONSOLE */
+	*cnt = 0;
+	return 0;
+#endif /* CONFIG_LIBUKCONSOLE */
```
`#endif /* !CONFIG_LIBUKCONSOLE */`

michpappas commented on lib/nolibc/stdio.c:
```diff
@@ -491,19 +496,19 @@ int fflush(FILE *fp __unused)
-int fputc(int _c, FILE *fp)
+int fputc(int _c __maybe_unused, FILE *fp __maybe_unused)
 {
+#ifdef CONFIG_LIBUKCONSOLE
```
`#if`

michpappas commented on lib/nolibc/stdio.c:
```diff
@@ -537,6 +543,9 @@ fputs_internal(const char *restrict s, FILE *restrict stream, int newline)
 	else
 		return 1;
+#else /* !CONFIG_LIBUKCONSOLE */
+	return EOF;
```
If the console is not enabled, perhaps we shouldn't be returning an error here but emulate success by returning `len` instead.

The same applies to the rest of the functions updated in this commit.

michpappas commented on lib/syscall_shim/Config.uk:
```diff
@@ -66,6 +66,7 @@ if LIBSYSCALL_SHIM
 		depends on LIBSYSCALL_SHIM_HANDLER
 			bool "'strace'-like messages for binary system calls"
 			default n
+			select LIBUKDEBUG if !LIBUKCONSOLE
```
Where does the change to use `libukdebug` derive from?

thass0 replied:
This is because we call into `ukdebug` if `LIBUKCONSOLE` isn't enabled. See the change in `lib/syscall_shim/uk_syscall_binary.c`. It's funny that we fall back on `ukdebug` for printing if `ukconsole` is missing because `ukdebug` just calls into `ukconsole` internally, so either way if `ukconsole` is disabled, we don't get output. But of course it's not guaranteed that `ukdebug` uses `ukconsole` to print stuff.

thass0 replied:
I think I discussed this with @skuenzer before

michpappas replied:
Okay, let's keep it for now then.

michpappas commented on lib/syscall_shim/uk_syscall_binary.c:
```diff
@@ -131,10 +135,16 @@ void ukplat_syscall_handler(struct uk_syscall_ctx *usc)
 	 */
-	ukplat_coutk(prsyscallbuf, (__sz) prsyscalllen);
+#if CONFIG_LIBUKCONSOLE
+	uk_console_out(prsyscallbuf, (__sz) prsyscalllen);
```
I'm skeptical towards this change, as one using `uk_console_` API prevents one to disable these messages via the relevant verbosity options.

thass0 replied:
I think these messages are supposed to mimic the output of `strace`. And that's why the extra information printed by `uk_pr_debug` and friends shouldn't be included.

michpappas commented on lib/posix-tty/Config.uk:
```diff
@@ -7,6 +7,7 @@ if LIBPOSIX_TTY
 
 	config LIBPOSIX_TTY_SERIAL
 	bool "Provide file driver for the serial console"
+	select LIBUKCONSOLE
```
Is this commit message accidentally left there?
```
    I forgot the update these calls because I started work on `ukconsole
    on an older branch that didn't have `posix-tty`.
```

michpappas commented on lib/ukdebug/print.c:
```diff
@@ -93,24 +96,34 @@
 	int prevlvl;
 };
 
+#ifndef CONFIG_LIBUKCONSOLE
```
`#if !CONFIG_LIBUKCONSOLE`

michpappas commented on lib/ukdebug/print.c:
```diff
@@ -93,24 +96,34 @@
 #define KLVL_DEBUG (-1)
 
-typedef int (*_ukplat_cout_t)(const char *, unsigned int);
+typedef __ssz (*_console_out_t)(const char *, __sz);
```
Since you're renaming this you can probably rename it to `uk_debug_console_out_func` as we discourage identifiers that start with an underscore.

michpappas commented on lib/ukdebug/print.c:
```diff
@@ -93,24 +96,34 @@
+/* This is the least invasive method of turning the `cout` call into a no-op
+ * if `ukconsole` isn't configured.
+ */
+__ssz uk_console_out(const char *, __sz)
```
IMO it looks a bit out of place to provide a definition in the scope of `libukconsole` here. An alternative way could be something along the lines of:
```c
static uk_debug_cout_func uk_debug_cout;

#if CONFIG_LIBUKCONSOLE
uk_debug_cout = uk_console_out;
#else /* CONFIG_LIBUKCONSOLE */
static __ssz uk_debug_cout_none(const char *buf, __sz len)
{
	return sz;
}
uk_debug_cout = uk_debug_cout_none;
#endif /* !CONFIG_LIBUKCONSOLE */

static struct _vprint_console kern  = { .cout = uk_debug_cout,
					.newline = 1,
					.prevlvl = INT_MIN };
```

thass0 replied:
The suggested solution doesn't compile unfortunately (`uk_debug_cout` isn't a compile-time constant). 

But we can do this:
```c
#if CONFIG_LIBUKCONSOLE
#define UK_DEBUG_COUT uk_console_out
#else
__ssz uk_debug_cout_none(const char *buf __unused, __sz len __unused)
{
	return 0;
}
#define UK_DEBUG_COUT uk_debug_cout_none
#endif /* !CONFIG_LIBUKCONSOLE */

/* Console state for kernel output */
#if CONFIG_LIBUKDEBUG_REDIR_PRINTD || CONFIG_LIBUKDEBUG_PRINTK
static struct _vprint_console kern  = { .cout = UK_DEBUG_COUT,
					.newline = 1,
					.prevlvl = INT_MIN };
#endif
```

michpappas replied:
Another option is:
```
#if CONFIG_LIBUKCONSOLE
#define UK_DEBUG_COUT        uk_console_cout
#else /* CONFIG_LIBUKCONSOLE */
#define UK_DEBUG_COUT        NULL
#endif /* CONFIG_LIBUKCONSOLE */
```
and then make all calls to `cout` conditional.

thass0 replied:
I'm resolving this as it's a duplicate of this comment: https://github.com/unikraft/unikraft/pull/1464#discussion_r1703655500

michpappas commented on drivers/uktty/com/Config.uk:
```diff
@@ -0,0 +1,69 @@
+menuconfig LIBUKTTY_COM
+	bool "COM serial port drivers"
+	default n
```
`default n` is redundant, elsewhere in this file too

michpappas commented on drivers/uktty/com/com.c:
```diff
@@ -25,47 +25,175 @@
+#include <uk/boot/earlytab.h>
 
-void _libkvmplat_init_serial_console(void)
+#define COM1_BASE 0x3f8
```
Same rules for macro definitions describe in other comments (align values with tabs).

michpappas commented on drivers/uktty/com/com.c:
```diff
@@ -25,47 +25,175 @@
-	if (a == '\n')
-		serial_write('\r');
-	serial_write(a);
+	return (int)inb(COM_STATUS(port_addr)) & 0x20;
```
Please use definitions instead of magic numbers. Ditto for `check_rx_ready()`.

michpappas commented on drivers/uktty/com/com.c:
```diff
@@ -25,47 +25,175 @@
+	return (int)inb(COM_STATUS(port_addr)) & 0x01;
 }
+
+/**
```
I'm not really keen of macro-generated code. What we should do instead is to rework these functions to be able derive the device. What you can probably do is:
* Define driver-specific structures that wrap around `struct uk_console` and add private data like port / MMIO base.
* Add the device parameter to ops
* Have the driver derive its wrapping struct via `container_of`.
* Rework the implementation of drivers to use the port / MMIO base.

michpappas commented on drivers/uktty/com/Makefile.uk:
```diff
@@ -0,0 +1,4 @@
+$(eval $(call addlib_s,libuktty_com,$(CONFIG_LIBUKTTY_COM)))
+
+LIBUKTTY_COM_CINCLUDES-y += -I$(UK_PLAT_COMMON_BASE)/include
+LIBUKTTY_COM_SRCS-y += $(LIBUKTTY_COM_BASE)/com.c|isr
```
Same question as before about using the `isr` flag.

michpappas commented on drivers/uktty/vga/Config.uk:
```diff
@@ -0,0 +1,5 @@
+menuconfig LIBUKTTY_VGA
+	bool "VGA console driver"
+	default n
```
Redundant

michpappas commented on drivers/uktty/vga/Makefile.uk:
```diff
@@ -0,0 +1,3 @@
+$(eval $(call addlib_s,libuktty_vga,$(CONFIG_LIBUKTTY_VGA)))
+
+LIBUKTTY_VGA_SRCS-y += $(LIBUKTTY_VGA_BASE)/vga.c|isr
```
Why `isr`?

skuenzer replied:
I think this is related to dicsussion https://github.com/unikraft/unikraft/pull/1464#discussion_r1699592876

michpappas commented on drivers/uktty/vga/exportsyms.uk:
```diff
@@ -0,0 +1 @@
+vga_early_init
```
This should not be exported if the driver is registered to earlytab

michpappas commented on plat/kvm/x86/console.c:
```diff
@@ -28,18 +28,12 @@
 #include <uk/plat/console.h>
 #include <uk/config.h>
 #include <uk/essentials.h>
-#if (CONFIG_KVM_DEBUG_VGA_CONSOLE || CONFIG_KVM_KERNEL_VGA_CONSOLE)
```
This is not necessary as you remove that file in a later commit anyway.

michpappas commented on drivers/uktty/pl011/Config.uk:
```diff
@@ -13,4 +14,13 @@ config LIBUKTTY_PL011_EARLY_CONSOLE_BASE
 
 endif
+
+config LIBUKTTY_PL011_OUTPUT
```
Given that we have one early console device, I think these essentially overlap with the `Early Console` option. Same for `ns16550`.

michpappas reviewed (changes requested):
Hi @thass0, please see the comments below 🚀

michpappas commented on drivers/uktty/pl011/Config.uk:
```diff
@@ -2,6 +2,8 @@ menuconfig LIBUKTTY_PL011
 	depends on ARCH_ARM_64
 	select LIBUKOFW
+	select LIBUKCONSOLE
+	select LIBUKALLOC
```
The idea is that ukalloc is optional, as console devices may be used in safety critical systems where dynamic allocation is a no-go. So what I would do is remove the dependency from Config.uk, and make any code that relies on dynamic allocation conditional to `CONFIG_LIBUKALLOC` instead.

michpappas commented on drivers/uktty/pl011/Makefile.uk:
```diff
@@ -1,10 +1,7 @@
 LIBUKTTY_PL011_CINCLUDES-y  += -I$(UK_PLAT_COMMON_BASE)/include
 
-LIBUKTTY_PL011_SRCS-y += $(LIBUKTTY_PL011_BASE)/pl011.c
+LIBUKTTY_PL011_SRCS-y += $(LIBUKTTY_PL011_BASE)/pl011.c|isr
```
TBD

michpappas commented on drivers/uktty/pl011/pl011.c:
```diff
@@ -25,7 +25,9 @@
-#include <uk/plat/console.h>
+#include <uk/console.h>
+#include <uk/compiler.h>
+#include <uk/alloc.h>
```
Conditional to `CONFIG_LIBUKALLOC`

michpappas commented on drivers/uktty/pl011/pl011.c:
```diff
@@ -79,45 +81,99 @@ static const char * const fdt_compatible[] = {
+};
+
+#if CONFIG_LIBUKTTY_PL011_EARLY_CONSOLE
+static struct pl011_device early_dev;
```
`earlycon` is a nice name :)

thass0 replied:
Agree 😊

michpappas commented on drivers/uktty/pl011/pl011.c:
```diff
@@ -79,45 +81,99 @@ static const char * const fdt_compatible[] = {
+
+static __ssz pl011_out(struct uk_console *dev, const char *buf, __sz len)
+{
+	struct pl011_device *pl011_dev =
```
This looks a bit ugly, and is generally discouraged by the latest coding conventions. I would change it to
```c
	struct pl011_device *pl011_dev;
	__sz l = len;

	pl011_dev = __containerof(dev, struct pl011_device, dev);
```

With the opportunity you can add some assertions for pointer parameters.

michpappas commented on drivers/uktty/pl011/pl011.c:
```diff
@@ -79,45 +81,99 @@ static const char * const fdt_compatible[] = {
+
+static __ssz pl011_in(struct uk_console *dev, char *buf, __sz len)
+{
+	struct pl011_device *pl011_dev =
```
This looks a bit ugly, and is generally discouraged by the latest coding conventions. I would change it to
```c
	struct pl011_device *pl011_dev;
	int rc;

	pl011_dev = __containerof(dev, struct pl011_device, dev);
```

With the opportunity you can add some assertions for pointer parameters.

michpappas commented on drivers/uktty/pl011/pl011.c:
```diff
@@ -169,28 +225,30 @@ static int early_init(struct ukplat_bootinfo *bi)
 
-	mrd.pbase = pl011_base;
-	mrd.vbase = pl011_base;
+	mrd.pbase = early_dev.base;
```
While you're here you can also add a comment I omitted before:
```
/* Add an mrd to keep the device mapped past init */
```

michpappas commented on drivers/uktty/pl011/pl011.c:
```diff
@@ -213,116 +271,66 @@ static int init(struct uk_init_ctx *ictx __unused)
+	a = uk_alloc_get_default();
+	UK_ASSERT(a);
+
+	do {
```
That's a pretty big block. What I would do is move the fdt part into a separate function conditional to `CONFIG_LIBUKALLOC`, and update init to:
```c
static int init(struct uk_init_ctx *ictx __unused)
{
#if CONFIG_LIBUKALLOC
	struct ukplat_bootinfo *bi;
	struct uk_alloc *a;
	int offset = -1;

	a = uk_alloc_get_default();
	UK_ASSERT(a);

	bi = ukplat_bootinfo_get();
	UK_ASSERT(bi);

	while(1) {
		pl011_dev = fdt_get_device((void *)bi->dtb, offset, a);
		if (!pl011_dev)
			break; /* no more devices */

		if (unlikely(PTRISERR(pl011_dev))) {
			uk_pr_err();
			return ERR2PTR(dev);
		}
#if CONFIG_PAGING
		pl011_dev->base = uk_bus_pf_devmap(pl011_dev->base, pl011_dev->size);
		if (unlikely(PTRISERR(pl011_dev->base))) {
			uk_pr_err("Could not map pl011\n");
			return PTR2ERR(pl011_dev->base);
		}
#endif /* !CONFIG_PAGING */

		rc = pl011_setup(pl011_dev->base);
		if (unlikely(rc)) {
			uk_pr_err("Could not initialize pl011\n");
			return rc;
		}

		uk_console_register(&pl011_dev->dev);
		uk_pr_info("tty: pl011 (%p)\n", &pl011_dev->base);
	}
#endif /* CONFIG_LIBUKALLOC */

	return 0;
}
```
That way you achieve that:
1.Different device detection methods can be easily added.
2. Systems that don't use dynamic allocation can still use the statically allocated earlycon, while more complex systems can register multiple devices.

michpappas commented on drivers/uktty/pl011/pl011.c:
```diff
@@ -213,116 +271,66 @@ static int init(struct uk_init_ctx *ictx __unused)
+	UK_ASSERT(a);
+
+	do {
+		/* Pick the 1st device in the dtb */
```
This is no longer the case as this will pick the next device from offset. You can drop this comment altogether.

michpappas commented on lib/ukconsole/include/uk/console.h:
```diff
@@ -0,0 +1,134 @@
+	struct uk_list_head _list;
+};
+
+#define UK_CONSOLE_OPS(_out, _in) \
```
I understand the reasoning for the `UK_CONSOLE` macro, but is there a reason for `UK_CONSOLE_OPS`?

thass0 replied:
There is no reason for `UK_CONSOLE_OPS` beyond staying consistent with `UK_CONSOLE`. I'm removing `UK_CONSOLE_OPS`.

michpappas commented on lib/nolibc/stdio.c:
```diff
@@ -446,14 +450,15 @@ int vfprintf(FILE *fp, const char *fmt, va_list ap)
-	else if (fp == stderr)
-		ret = ukplat_coutd(buf, ret);
+	if (fp == stdout || fp == stderr)
+		ret = uk_console_out(buf, ret);
```
What I would do instead is let vsnprintf() above do the formatting so that we can return a reasonable value even if `libukconsole` is not enabled, and then make the printing conditional to `libukconsole`:
```
        if (fp == stdout || fp == stderr)
#if CONFIG_LIBUKCONSOLE
                ret = ukplat_coutd(buf, ret);
#else /* CONFIG_LIBUKCONSOLE */
                ret = strlen(buf);
#endif /* CONFIG_LIBUKCONSOLE */
```

thass0 replied:
Makes sense, I like the idea

michpappas commented on lib/nolibc/stdio.c:
```diff
@@ -491,20 +496,20 @@ int fflush(FILE *fp __unused)
 	if (ret == 1)
 		return _c;
 
-	return EOF;
```
Here you lost the `return EOF;` if `libukconsole` is not enabled.

michpappas commented on lib/devfs/Config.uk:
```diff
@@ -29,4 +29,5 @@ if LIBDEVFS
 		bool "Register stdout device"
 		default y if LIBDEVFS_AUTOMOUNT
 		default n
```
You can get rid of that redundant `default n` while you're there.

thass0 replied:
Just that one or all of them? Can't tell from the outside if the existing `default n`s in this file are redundant. (I know `default n` is redundant by default, but there are edge-cases, right?)

michpappas replied:
You're right, I didn't notice there are more. In that case just leave them as is, as changing them all gets a out of scope.

michpappas commented on drivers/uktty/com/Config.uk:
```diff
@@ -0,0 +1,73 @@
+menuconfig LIBUKTTY_COM4
+	bool "Enable COM4 driver (0x2e8)"
+
+choice
```
Ideally this should be a choice for every individual COM.

michpappas commented on drivers/uktty/com/exportsyms.uk:
```diff
@@ -0,0 +1,4 @@
+com_com1_init
```
I think these can now go away

thass0 replied:
Indeed

michpappas commented on plat/common/include/uk/plat/common/memory.h:
```diff
@@ -157,12 +157,12 @@ ukplat_memregion_list_insert_legacy_hi_mem(struct ukplat_memregion_list *list)
 				.len = X86_VIDEO_MEM_LEN,
 				.pg_count = PAGE_COUNT(X86_VIDEO_MEM_LEN),
```
On the commit message:
> * The delete console driver configuration options now
>  live in the `drivers/uktty/` subfolders

Folders are for windows people, *nix hackers use directories 😛
https://en.wikipedia.org/wiki/Directory_(computing)#Folder_metaphor

thass0 replied:
Changing this immediately 🫡 We don't do folders here

michpappas commented on plat/common/include/uk/plat/common/memory.h:
```diff
@@ -157,12 +157,12 @@ ukplat_memregion_list_insert_legacy_hi_mem(struct ukplat_memregion_list *list)
 				.pg_count = PAGE_COUNT(X86_VIDEO_MEM_LEN),
 				.type  = UKPLAT_MEMRT_RESERVED,
-#if (CONFIG_KVM_DEBUG_VGA_CONSOLE || CONFIG_KVM_KERNEL_VGA_CONSOLE)
+#if CONFIG_LIBUKTTY_VGA
```
Although much better than before, I still don't like how a driver parameter creeps in here. Perhaps we can look up the himem mrd and append the write protection in the vga driver.

Update: Actually you can actually move the insertion of the `X86_VIDEO_MEM` mrd to the VGA driver 😉

thass0 replied:
I moved the insertion of the `X86_VIDEO_MEM` MRD to the VGA driver and deleted the code in `memory.h`

michpappas commented on lib/ukconsole/console.c:
```diff
@@ -0,0 +1,101 @@
+#include <uk/console.h>
+
+/* List of dynamically registered devices */
+UK_LIST_HEAD(uk_console_device_list);
```
Should be `static`

michpappas commented on drivers/uktty/vga/vga.c:
```diff
@@ -229,3 +197,64 @@ void _libkvmplat_vga_putc(char c)
+static struct uk_console vga_dev = UK_CONSOLE("VGA", &vga_ops,
+					      UK_CONSOLE_FLAG_STDOUT);
+
+int vga_init(struct ukplat_bootinfo *bi __unused)
```
Needs to be `static`. Also the ones above.

michpappas reviewed (commented):
@thass0 thanks for the updates, here's another round

michpappas commented on lib/nolibc/scanf.c:
```diff
@@ -49,6 +54,10 @@ uk_scanf(void *buffer,  size_t *cnt)
 	return 0;
+#else /* !CONFIG_LIBUKCONSOLE */
+	*cnt = 0;
+	return 0;
```
There is a fine difference between zero and `EOF`. According to scanf(3):
```
On success, these functions return the number of input items successfully matched and assigned; this can be fewer than provided for, or even zero, in the event of an early matching failure.

The value EOF is returned if the end of input is reached before either the first successful conversion or a matching failure occurs.  EOF is also returned if a read error occurs, in which case the error indicator for the stream (see ferror(3)) is set, and errno is set to indicate the error.
```

In other words, an application would interpret a zero as a value that didn't match the pattern. I wonder if we should be returning `EOF` here instead, in case the application handles these cases differently. Unlike outputting where we can emulate successful behavior, perhaps it's reasonable to assume that one can't expect input without enabling `libukconsole`?

thass0 replied:
Yes, we can't emulate successful behavior if we are doing input. I'm changing the return value to `EOF`.

michpappas commented on lib/ukdebug/print.c:
```diff
@@ -93,24 +96,35 @@
 #define KLVL_DEBUG (-1)
 
-typedef int (*_ukplat_cout_t)(const char *, unsigned int);
+typedef __ssz (*uk_debug_cout_func)(const char *buf, __sz len);
```
Actually you can simplify naming to `cout_func`, as typedefs don't need to be namespaced (they define types, not instances). Then the structure declaration below will look cleaner:
```diff
struct _vprint_console {
-	uk_debug_cout_func cout;
+	cout_func cout;
	int newline;
	int prevlvl;
};
```

michpappas commented on drivers/uktty/vga/Makefile.uk:
```diff
@@ -1,3 +1,6 @@
 $(eval $(call addlib_s,libuktty_vga,$(CONFIG_LIBUKTTY_VGA)))
 
+# TODO: Remove as soons as plat dependencies go away
```
"as soon as"

michpappas commented on drivers/uktty/com/com.c:
```diff
@@ -25,47 +25,291 @@
+		return rc;
+#endif /* CONFIG_LIBUKTTY_COM2 && !CONFIG_LIBUKTTY_COM2_EARLY */
+#if CONFIG_LIBUKTTY_COM3 && !CONFIG_LIBUKTTY_COM3_EARLY
+	rc = com_init_port(a, "COM2", COM2_PORT,
```
COM3

michpappas replied:
The parameter to `com_init_port()` for `COM3` is now `COM4`...

michpappas commented on drivers/uktty/com/com.c:
```diff
@@ -25,47 +25,291 @@
+		return rc;
+#endif /* CONFIG_LIBUKTTY_COM3 && !CONFIG_LIBUKTTY_COM3_EARLY */
+#if CONFIG_LIBUKTTY_COM4 && !CONFIG_LIBUKTTY_COM4_EARLY
+	rc = com_init_port(a, "COM2", COM2_PORT,
```
COM4

michpappas commented on lib/ukdebug/print.c:
```diff
@@ -41,6 +41,9 @@
 
 #include <uk/essentials.h>
 #include <uk/plat/console.h>
+#ifdef CONFIG_LIBUKCONSOLE
```
`#if`

michpappas reviewed (changes requested):
Hi @thass0, I did yet another round.

michpappas commented on lib/ukconsole/console.c:
```diff
@@ -0,0 +1,113 @@
+	if (!len)
+		return 0;
+
+	if (!buf)
```
```c
if unlikely(!buf)
```
Also in error conditions you should return values from errno, i.e. in this case `-EINVAL`.

michpappas commented on lib/ukconsole/console.c:
```diff
@@ -0,0 +1,113 @@
+{
+	struct uk_console *dev = __NULL;
+
+	if (!len)
```
```c
if (unlikely(!len))
```

michpappas commented on lib/ukconsole/include/uk/console.h:
```diff
@@ -0,0 +1,172 @@
+static inline __ssz uk_console_out_with(struct uk_console *dev,
+					const char *buf, __sz len)
+{
+	if (!len)
```
```c
	if (unlikely(!len))
		return 0;

	if (unlikely(!buf || !dev || !dev->ops || !dev->ops->out))
		return -EINVAL;
```

michpappas commented on lib/ukconsole/include/uk/console.h:
```diff
@@ -0,0 +1,172 @@
+static inline __ssz uk_console_in_with(struct uk_console *dev,
+				       char *buf, __sz len)
+{
+	if (!len)
```
```c
	if (unlikely(!len))
		return 0;

	if (unlikely(!buf || !dev || !dev->ops || !dev->ops->in))
		return -EINVAL;
```
Be careful with copy-paste, this was checking against `ops->out`.

michpappas commented on lib/ukconsole/console.c:
```diff
@@ -0,0 +1,113 @@
+	__sz leftover = len;
+	__ssz rc = 0;
+
+	if (!len)
```
```c
if (unlikely(!len))
	return 0;

if (unlikely(!buf))
	return -EINVAL;
```

michpappas commented on lib/ukdebug/print.c:
```diff
@@ -93,28 +96,47 @@
-					.newline = 1,
-					.prevlvl = INT_MIN };
+static struct _vprint_console debug = {
+#if CONFIG_LIBUKCONSOLE
```
IMO this looks a bit ugly, it would be cleaner to:
```c
#if CONFIG_LIBUKCONSOLE
#define UK_COUT_FUNC        uk_console_out
#else /* CONFIG_LIBUKCONSOLE */
#define UK_COUT_FUNC        NULL
#endif /* CONFIG_LIBUKCONSOLE */
```

and then define your vprint stuctures as:
```c
static struct _vprint_console kern  = {
	.cout = UK_COUT_FUNC,
	.newline = 1,
	.prevlvl = INT_MIN
};

static struct _vprint_console debug = {
	.cout = UK_COUT_FUNC,
	.newline = 1,
	.prevlvl = INT_MIN
};
```

thass0 replied:
I opted for the current option because it localized the change to the struct initialization and makes it more apparent what's going on.

thass0 replied:
Can we resolve this?

michpappas replied:
Yes, it's okay to resolve 👍🏼

procub3r commented:
I was referring to `drivers/uktty/vga` to write a GOP driver (`drivers/uktty/gop`) when I noticed that the VGA driver does not parse ANSI escape codes to format its output. Is this intended behavior?

When the `CONFIG_LIBUKDEBUG_ANSI_COLOR` option is disabled, everything is fine:
![image](https://github.com/user-attachments/assets/507a1a7b-b1ce-4b88-a1c3-9b8bd835e0b5)


When the `CONFIG_LIBUKDEBUG_ANSI_COLOR` option is enabled, the VGA output looks like this:
![image](https://github.com/user-attachments/assets/af84b471-bd0f-4a15-909f-119efb737934)

Expected output: The ANSI escape codes must be parsed and the output must be appropriately formatted (fg color, bg color etc).

CC: @mogasergiu

thass0 commented:
Seems like a valid point. I checked the history and it looks like the VGA driver was just forgotten when colors were added in commit edd3a6f0bf1e8644f9af9f67238ab69c0c762014. This commit is from 2020. There haven't been significant changes to `plat/kvm/x86/vga_console.c` since 2018.

procub3r commented:
I've implemented an ANSI escape code parser for the GOP driver. I bet it can be reused for the VGA driver as well. I could look into this in the near future if that's alright :grin:

thass0 commented:
Totally fine with me. Any improvements to the driver would be a win. Don't know if anybody else has an opinion on this

skuenzer commented:
> I've implemented an ANSI escape code parser for the GOP driver. I bet it can be reused for the VGA driver as well. I could look into this in the near future if that's alright :grin:

In fact, we do not have ANSI parsing for VGA yet. It would be a very nice addition if you could open another PR that adds it to the VGA driver. 👍

skuenzer commented on drivers/uktty/com/Config.uk:
```diff
@@ -0,0 +1,131 @@
+
+config LIBUKTTY_COM1_EARLY
+    bool "COM1"
+    imply LIBUKTTY_COM1
```
I think it is better to do a `depends on` instead of `imply`. Problem right now: You can configure a COM port as early console and then unselect the COM port driver which I would consider broken configuration state. I first thought about `select` but `depends on` would allow us to select the first available COM device as early console as default. As soon the driver gets unselected, it should automatically jump back to `None`. With this, a platform could just enable early console initialization and the enabled COM port (by the platform) with the lowest number is selected automatically on a fresh config. Unfortunately, a platform cannot set a choice option. Only boolean options can be activated (set to true) via other KConfig options (`select`, `imply`).

I am thinking of this (untested):
```
menuconfig LIBUKTTY_COM1
	bool "Enable COM1 driver (0x3f8)"
# <...>

menuconfig LIBUKTTY_COM2
	bool "Enable COM2 driver (0x2f8)"
# <...>

config LIBUKTTY_COM_EARLY_CONSOLE
    bool "Initialize early console"
    depends on LIBUKTTY_COM1_EARLY || LIBUKTTY_COM2_EARLY || ...

choice
    prompt "Console device"
    default LIBUKTTY_COM4_EARLY if !LIBUKTTY_COM1 && !LIBUKTTY_COM2 && !LIBUKTTY_COM3 && LIBUKTTY_COM4
    default LIBUKTTY_COM3_EARLY if !LIBUKTTY_COM1 && !LIBUKTTY_COM2 && LIBUKTTY_COM3
    default LIBUKTTY_COM2_EARLY if !LIBUKTTY_COM1 && LIBUKTTY_COM2
    default LIBUKTTY_COM1_EARLY
    depends on LIBUKTTY_COM_EARLY_CONSOLE # This should deprecate the "None" option

config LIBUKTTY_COM1_EARLY
    bool "COM1"
    depends on LIBUKTTY_COM1
    
config LIBUKTTY_COM2_EARLY
    bool "COM2"
    depends on LIBUKTTY_COM2

config LIBUKTTY_COM3_EARLY
    bool "COM3"
    depends on LIBUKTTY_COM3

config LIBUKTTY_COM4_EARLY
    bool "COM4"
    depends on LIBUKTTY_COM4
```

skuenzer replied:
I saw that the driver still compiles if all console devices are unselected: very good because we want to avoid that any possible configuration causes a broken build. I think it is good this way and we do not need to automatically confiugre a default device because a platform's `Config.uk` would do that.

skuenzer commented on lib/ukconsole/console.c:
```diff
@@ -0,0 +1,113 @@
+	UK_ASSERT(dev->ops);
+
+	uk_list_for_each_entry(known_dev, &uk_console_device_list, _list) {
+		if (dev == known_dev)
```
This is to double check that we are not registering a device twice? If so, I think, I'd prefer using an `UK_ASSERT` because I would consider it as API misuse. I would also guard the list iteration so that it gets compiled in only if assertions are enabled. What do you think?

skuenzer commented on lib/ukconsole/console.c:
```diff
@@ -0,0 +1,113 @@
+	 * there has not yet been another device with any flags set, we give
+	 * the current device flags. Now we have at least one device with flags
+	 */
+	if (!uk_console_set_standard_once && !dev->flags) {
```
For being future proof in respect of new flags, what do you think of doing this:
```diff
-!dev->flags
+!(dev->flags & UK_CONSOLE_FLAG_STDOUT)
```

skuenzer commented on lib/ukconsole/console.c:
```diff
@@ -0,0 +1,113 @@
+	}
+
+	uk_list_add_tail(&dev->_list, &uk_console_device_list);
+	dev->id = uk_console_device_count++;
```
Probably we should emit a log message that a new device got registered (`uk_pr_info`). At least a `TODO` as placeholder for the future would make sense. I am thinking that as soon as we have a kernel message buffer, it is very useful for debugging purposes to have a message recorded as soon as a device got registered and which flags it (got) set:

```c
 uk_pr_info("Registered console%"PRIu16": %p (%s), flags: %c%c\n",
            dev->id, dev, dev->name,
            (dev->flags & UK_CONSOLE_FLAG_STDIN) ? 'I' : '-',
            (dev->flags & UK_CONSOLE_FLAG_STDOUT) ? 'O' : '-',
 );
```

skuenzer commented on lib/ukconsole/include/uk/console.h:
```diff
@@ -0,0 +1,176 @@
+  *   - (>=0): Number of bytes written
+  *   - (<0): Error
+  */
+static inline __ssz uk_console_out_with(struct uk_console *dev,
```
I am not happy with the `_with` as suffix for these two functions. I think it reads harder because I expect a noun after `with`, like `with_device`. However, in this project, we prefer shorter naming, so, I am preferring to append only one word. Typically, adjectives or nouns are used in such cases. What do you think of `direct`, for example?:
```
uk_console_out_direct
uk_console_in_direct
```
`single`, `dedicated`, or `specific` may work, as well. I am open for any other suggestion.

thass0 replied:
I think `direct` is good.

skuenzer commented on lib/ukconsole/include/uk/console.h:
```diff
@@ -0,0 +1,176 @@
+ *
+ * @return Number of devices registered with `ukconsole`
+ */
+__u16 uk_console_get_device_count(void);
```
Can we align the naming of this interface with the ones in `uknetdev` or `ukblkdev`:
`uk_console_count()`?

skuenzer commented on lib/ukconsole/include/uk/console.h:
```diff
@@ -0,0 +1,176 @@
+	((struct uk_console) {			\
+		.id = 0,			\
+		.name = (_name),		\
+		.flags = _flags,		\
```
Why no parentheses here?

thass0 replied:
Forgot them, I suppose :)

skuenzer commented on lib/ukconsole/include/uk/console.h:
```diff
@@ -0,0 +1,176 @@
+
+#define UK_CONSOLE(_name, _ops, _flags)		\
+	((struct uk_console) {			\
+		.id = 0,			\
```
We can assume that the compiler will initialize not listed fields with `0` automatically (partial initialization), `.id` would not be needed.

skuenzer commented on lib/ukconsole/include/uk/console.h:
```diff
@@ -0,0 +1,176 @@
+	struct uk_list_head _list;
+};
+
+#define UK_CONSOLE(_name, _ops, _flags)		\
```
Hum... I would prefer naming this as `UK_CONSOLE_DECLARE()`

skuenzer commented on lib/ukconsole/include/uk/console.h:
```diff
@@ -0,0 +1,176 @@
+ * @param dev
+ *   The device driver to register
+ */
+void uk_console_register(struct uk_console *dev);
```
I think we should also here split the API into two categories:
- Device usage (`uk_console_in`, `uk_console_out` and co.)
- Console driver implementation: `uk_console_register`, the `struct`s definitions and datatypes only releavnt for a driver.

In the past we caused API confusion with uknetdev when we had these two categories declared in the same header file. We fixed it by moving the declarations into two separate headers following the just mentioned categories. For `ukconsole` this means:

* `uk/console.h`:
  * Anonymous device struct declaration (because more is not needed here yet): `struct uk_console;`
  * The following functions:
```c
 uk_console_count()
 uk_console_get()
 uk_console_out()
 uk_console_in()
 uk_console_out_direct()
 uk_console_in_direct()
```

* `uk/console/driver.h` (previously we had the convention `uk/console_driver.h`):
  * Exact struct defintion
  * Function pointer data types
  * Registration macros and functions: `uk_console_register()`
  * Helper macros: `UK_CONSOLE()`

skuenzer commented on lib/posix-tty/serial.c:
```diff
@@ -118,7 +118,7 @@ static ssize_t serial_write(const struct uk_file *f __maybe_unused,
 			return  -EFAULT;
 
-		bytes_written = ukplat_coutk(buf, len);
+		bytes_written = uk_console_out(buf, len);
```
I think we need this `\n` -> `\r\n` conversion in this file with this change.

thass0 replied:
This is done in the gdb stub PR. See https://github.com/unikraft/unikraft/pull/1479/commits/a5f770a8ae07845254b3d660802b377def99c81c

skuenzer replied:
👍🏻 Right, I left there (PR #1479) a comment about maybe migrating into this PR. But both is fine. We want to migrate both PRs.

skuenzer commented on drivers/uktty/com/com.c:
```diff
@@ -25,47 +25,290 @@
+#include <uk/alloc.h>
+#endif /* CONFIG_LIBUKALLOC */
+
+#if CONFIG_LIBUKLIBPARAM
```
This guard is not needed. lib/uklibparam provides a header with stubs as soon as removed from the build.

skuenzer commented on drivers/xen/console/Config.uk:
```diff
@@ -1,8 +1,23 @@
 		the emergency console
+
+if LIBXEN_CONSOLE
+config LIBXEN_CONSOLE_HV
```
Hum... I am thinking that we should change this to have two drivers:
One for the emergency console (just hypercalls) and another one for the standard Xen console. The emergency console (due to its simplicity) can also directly register as early console.
This driver split would make these sub configurations unnecessary.

skuenzer commented on drivers/uktty/Config.uk:
```diff
@@ -1,5 +1,6 @@
 menu "Serial console"
 
+source "$(KCONFIG_DRIV_BASE)/uktty/com/Config.uk"
```
Will we adopt the `uktty` folder name to `ukconsole` as soon we merge `com` and `ns16550`?

michpappas replied:
We can do it in this PR (separate commit please). On a similar note, we should probably also move the contents of the `drivers/uktty/com` into `drivers/uktty/ns16550`, as these are essentially the same driver.

thass0 committed:
drivers: Rename `drivers/uktty` to `drivers/ukconsole`

The `drivers/ukconsole` subdirectory is where drivers implementing
the interface from the `ukconsole` library will live.

Signed-off-by: Thassilo Schulze <thassilo@unikraft.io>

thass0 committed:
drivers/xen/emg_console: Add Xen emergency console

The emergency console is kept in a separate library from
the HV console for simplicity. The emergency console also
registers in the early init tab.

Signed-off-by: Thassilo Schulze <thassilo@unikraft.io>

michpappas commented on drivers/ukconsole/Config.uk:
```diff
@@ -0,0 +1,8 @@
+menu "Serial console"
```
With the opportunity of the rename from `uktty` to `ukconsole` you can also replace this `Config.uk` with an entry in `drivers/Config.uk` since that is the right way to register a driver subsystem:
```diff
+ menu "Serial Console"
+
+source "$(shell,$(UK_BASE)/support/build/config-submenu.sh -q -o '$(KCONFIG_DIR)/drivers-console.uk' -r '$(KCONFIG_DRIV_BASE)/ukconsole' -l '$(KCONFIG_DRIV_BASE)/ukconsole' -e '$(KCONFIG_EXCLUDEDIRS)')"
+
+endmenu
```

skuenzer replied:
Note that `drivers/ukconsole/Config.uk` must not be provided if you use the `config-submenu.sh` script in `drivers/Config.uk`. This one generates the file on-the-fly.

michpappas commented on drivers/ukconsole/ns16550/Makefile.uk:
```diff
@@ -0,0 +1,10 @@
+$(eval $(call addplatlib_s,kvm,libukconsole_ns16550,$(CONFIG_LIBUKCONSOLE_NS16550)))
```
The adaptations of drivers for the rename from `uktty` to `ukconsole` needs to happen in a single scope. The commit this chage is currently in ("Adapt NS16550 for use with `ukconsole`") contains changes that are irrelevant to the renaming. Same for the pl011 driver.

You have two options:
1. Make all rename-related changes in the commit where you do the directory rename.
2. Make one commit with the rename of the directory, followed by one commit with changes in `pl011`,followed by one commit with changes in `ns16550`.

Also in the commit message:
```
    These changes are functionally identical to those in the
    PL011 driver.
```
As a general rule, because commits are cherry-picked here and there, it's better to preferable to describe the changes made, even if that repeats the another commit message, so that the description always accompanies the changes regardless of the tree it ends up to.

thass0 replied:
I'm adding an extra commit just for the renames. Also updating the commit message for the NS16550 commit

michpappas commented on drivers/ukconsole/ns16550/ns16550.c:
```diff
@@ -253,143 +337,140 @@ static int early_init(struct ukplat_bootinfo *bi)
+	console_dev->size = dev->size;
+
+	uk_console_register(&console_dev->dev);
+	uk_pr_info("CONSOLE: ns16550 (%p)\n", &console_dev->base);
```
Why is it shouting "CONSOLE"? 😆 Or is this accidental search and replace?

thass0 replied:
yes, it it's accidental search and replace :joy:

michpappas commented on drivers/xen/console/include/uk/console/console.h:
```diff
@@ -34,17 +34,6 @@
 #ifndef __CONSOLE_H__
```
Similarly to virtio, xen drivers don't follow the usual per-subsystem directory layout, but rather act as a technology-specific collection. Because of that, the conventions are slightly different:

* The include guards should be renamed to `__XEN_CONSOLE_H__`.
* The path to this header should be `#include <xen/console.h>`. Or `<uk/xen/console.h> maybe @skuenzer has some input here. For virtio we use `<virtio/header.h>`, yet `drivers/xenbus/` uses `uk/xenbus/header.h>`.

For similar cases see also the `drivers/virtio`.

skuenzer commented on drivers/xen/emg_console/emg_console.c:
```diff
@@ -71,18 +71,42 @@
-#include "emg_console.h"
-
-int emg_console_output(const char *str, unsigned int len)
+__ssz console_out(struct uk_console *dev __unused, const char *str, __sz len)
```
`static`?

skuenzer replied:
`static`?

skuenzer commented on drivers/xen/emg_console/emg_console.c:
```diff
@@ -71,18 +71,42 @@
 	return len;
 }
+
+__ssz console_in(struct uk_console *dev __unused, char *str __unused,
```
`static`?

skuenzer replied:
I am also wondering if we should support `__NULL` as out or in operation or if we should provide a generic stub one that can be used and is provided by `#include <uk/console/driver.h>`.

thass0 replied:
`ukconsole` is written so that a `__NULL` op is safe as long as the flag to use that operation isn't set. But I think drivers should generally provide both functions.

thass0 replied:
In fact, I'd rather forbid setting an op to `__NULL` altogether.

skuenzer replied:
As we discussed offline, there is a good reason to support NULL as in and out function. It allows us to improving auto-picking the first console input/output device during boot since we can detect device support with `NULL` pointers in the ops struct.

michpappas commented on drivers/xen/console/Config.uk:
```diff
@@ -0,0 +1,7 @@
+menuconfig LIBXEN_CONSOLE
```
Please indent with tabs

michpappas commented on drivers/xen/emg_console/Config.uk:
```diff
@@ -0,0 +1,6 @@
+menuconfig LIBXEN_EMG_CONSOLE
```
Please indent with tabs

michpappas commented on drivers/xen/emg_console/emg_console.c:
```diff
@@ -71,18 +71,42 @@
 	return len;
 }
+
+__ssz console_in(struct uk_console *dev __unused, char *str __unused,
```
`console_in` and `console_out` should be `static`?

michpappas commented on drivers/xen/console/console.c:
```diff
@@ -254,4 +239,59 @@ void hv_console_init(void)
+	return 0;
+}
+
+__ssz console_out(struct uk_console *dev __unused, const char *str, __sz len)
```
All private functions should be `static`.

michpappas commented on drivers/xen/console/console.c:
```diff
@@ -254,4 +239,59 @@ void hv_console_init(void)
+}
+#endif
+
+UK_BOOT_EARLYTAB_ENTRY(hv_console_prepare, UK_PRIO_EARLIEST);
```
Just a note for the future, that anything registered to `UK_PRIO_EARLIEST` is not guaranteed to have access to the cmdline as that is also parsed at the same priority level. Also this function depends on `events_init()` which should eventually also move to the earlytab before `hv_console_prepare()`, and I'm not sure whether these also have dependencies to traps or shared_info. So for now it's probably okay as is, but I anticipate that it will be rearranged later as more things move into earlytab.

thass0 replied:
Good point

skuenzer commented on lib/ukconsole/include/uk/console/driver.h:
```diff
@@ -0,0 +1,86 @@
+extern "C" {
+#endif
+
+struct uk_console;
```
You could also `#include <uk/console.h>` instead. Both works, actually.

thass0 replied:
I'll keep it as is

skuenzer commented on lib/ukconsole/include/uk/console/driver.h:
```diff
@@ -0,0 +1,86 @@
+static inline void uk_console_init(struct uk_console *dev, const char *name,
+				   struct uk_console_ops *ops, int flags)
+{
+	UK_ASSERT(dev);
```
We should probably also assert `ops` here already. I saw that you do that during register as well. I think at both places is a good idea for pin-pointing a broken driver.

skuenzer commented on lib/ukconsole/include/uk/console/driver.h:
```diff
@@ -0,0 +1,86 @@
+
+struct uk_console {
+	__u16 id;
+	struct uk_console_ops *ops;
```
Can be `const` right?

skuenzer commented on lib/ukconsole/include/uk/console/driver.h:
```diff
@@ -0,0 +1,86 @@
+ *   Requested flags for the device
+ */
+static inline void uk_console_init(struct uk_console *dev, const char *name,
+				   struct uk_console_ops *ops, int flags)
```
ops could be `const`

skuenzer commented on lib/ukconsole/include/uk/console/driver.h:
```diff
@@ -0,0 +1,86 @@
+	dev->name = name;
+	dev->ops = ops;
+	dev->flags = flags;
+	dev->id = 0;
```
If you do
```c
*dev = { .name = name; .ops = ops; .flags = flags, .id = 0 }
```
then the code would automatically zero out new fields that are added in the future.

skuenzer replied:
Btw, initializing `.id` with `__U16_MAX` can indicate that the `uk_console` struct was never registered and might be a useful detail for debugging.

michpappas commented on drivers/xen/emg_console/emg_console.c:
```diff
@@ -71,18 +71,42 @@
+	return 0;
+}
+
+UK_BOOT_EARLYTAB_ENTRY(init_console, UK_PRIO_EARLIEST);
```
Similar comments with the priorities of `console.c`. I expect that there are no dependencies with cmdline.

michpappas reviewed (changes requested):
@thass0 a few more comments

thass0 committed:
drivers/ukconsole: Rename all occurrences of `uktty` to `ukconsole`

This is part of the rename of the `drivers/uktty` subdirectory
to `drivers/ukconsole`.

Signed-off-by: Thassilo Schulze <thassilo@unikraft.io>

skuenzer commented on lib/ukconsole/tests/test_ukconsole.c:
```diff
@@ -0,0 +1,74 @@
+		== &uk_console_test_dev);
+
+	uk_console_out_direct(uk_console_get(uk_console_test_dev.id),
+			      "Hello out with!", 15);
```
I guess the string with `with` is a leftover. 😄 Not really important since it is a testcase only.

skuenzer commented on lib/ukdebug/print.c:
```diff
@@ -93,28 +97,47 @@
 };
 
 /* Console state for kernel output */
 #if CONFIG_LIBUKDEBUG_REDIR_PRINTD || CONFIG_LIBUKDEBUG_PRINTK
```
Hum, all these options regarding kernel console and debug console do not make sense anymore since we end up with just one console after this PR. So this code could be completely removed. However, this we can do in a follow-up PR.

skuenzer replied:
CC @michpappas

skuenzer commented on drivers/ukconsole/vga/Config.uk:
```diff
@@ -0,0 +1,4 @@
+menuconfig LIBUKCONSOLE_VGA
```
We can use `config` here because we do not have any driver configs

skuenzer commented on drivers/ukconsole/vga/Makefile.uk:
```diff
@@ -0,0 +1,3 @@
+$(eval $(call addlib_s,libukconsole_vga,$(CONFIG_LIBUKCONSOLE_VGA)))
+
+LIBUKCONSOLE_VGA_SRCS-y += $(LIBUKCONSOLE_VGA_BASE)/vga.c|isr
```
We may need to revisit this `isr` property. I saw that some drivers have it assigned some others do not have it. I think best is for now that we comple the whole gbstub code with `|isr`, also `ukconsole` and all console drivers. This should ensure that we can debug trap handlers as well. Best is to revisit optimizing this later because I expect that we will have integrated compiler support to check that isr functions are only calling other isr functions.

michpappas commented on drivers/ukconsole/ns16550/Config.uk:
```diff
@@ -0,0 +1,141 @@
+
+if ARCH_X86_64
+
+config LIBUKCONSOLE_COM_EARLY_CONSOLE
```
Regarding the last update that moves the `com` into `ns16550`. I think you need to update the config directives to match the rest of the driver, i.e. `CONFIG_UKCONSOLE_NS16550_...`. Indentation in `Config.uk` must use tabs.

Also remember to update the `Config.uk` of KVM, as currently no console drivers is selected for `x86_64`.

michpappas commented on drivers/ukconsole/ns16550/Config.uk:
```diff
@@ -0,0 +1,141 @@
+	bool "NS16550 / 8250"
+	depends on (ARCH_ARM_64 || ARCH_X86_64)
+	select LIBUKOFW if ARCH_ARM_64
+    select LIBUKCONSOLE
```
This needs a tab

skuenzer commented on lib/ukconsole/console.c:
```diff
@@ -0,0 +1,149 @@
+		return -EINVAL;
+
+	/* Collect input from all STDIN devices and append it */
+	uk_list_for_each_entry(dev, &uk_console_device_list, _list) {
```
Can you add the following note:
```c
/*
 * FIXME: This code works well if we do not expect input
 *        from multiple devices at the same time. Due to
 *        the restart of the iteration each time
 *        `uk_console_in()` is called, devices with lower
 *        registration IDs are given higher priority.
 *        We could solve this by remembering the iteration
 *        point between calls.
 */
```

skuenzer replied:
Because we are looking into `dev->ops->in` the following assertion would make sense on each iteration:
```c
UK_ASSERT(dev->ops);
```

skuenzer commented on lib/ukconsole/console.c:
```diff
@@ -0,0 +1,149 @@
+		return 0;
+
+	if (unlikely(!buf || !dev || !dev->ops || !dev->ops->in))
+		return -EINVAL;
```
`!buf || !dev || !dev->ops` should be catched by an assertion.
In case `!dev->ops->in`, I would return `-EIO`.

skuenzer replied:
👍🏻 We adopted as discussed offline

skuenzer commented on lib/ukconsole/console.c:
```diff
@@ -0,0 +1,149 @@
+		return 0;
+
+	if (unlikely(!buf || !dev || !dev->ops || !dev->ops->out))
+		return -EINVAL;
```
`!buf || !dev || !dev->ops` should be catched by an assertion.
In case `!dev->ops->out`, I would return `-EIO`.

skuenzer replied:
👍🏻 We adopted as discussed offline

skuenzer commented on lib/ukconsole/include/uk/console/driver.h:
```diff
@@ -0,0 +1,87 @@
+	dev->name = name;
+	dev->ops = ops;
+	dev->flags = flags;
+	dev->id = __U16_MAX;
```
If you do

```c
*dev = { .name = name; .ops = ops; .flags = flags; .id = __U16_MAX }
```

then the code would automatically zero out new fields that are added in the future.

skuenzer commented on lib/ukconsole/tests/test_ukconsole.c:
```diff
@@ -0,0 +1,75 @@
+
+	for (i = 0; i < dev_cnt; i++)
+		UK_TEST_EXPECT_NOT_NULL(uk_console_get(i));
+}
```
Another test after the loop could be:
```c
UK_TEST_EXPECT_NULL(uk_console_get(dev_cnt));
```

skuenzer commented on drivers/ukconsole/ns16550/Config.uk:
```diff
@@ -1,16 +1,141 @@
+
+if ARCH_X86_64
+
+config LIBUKCONSOLE_NS16550_COM_EARLY_CONSOLE
```
Due to the dependencies to each individual console, I would prefer if the early console setting and choice to appear after listing the COM ports. This would slightly make more sense UI-wise: Everything below a settings can live adapt to changes but nothing before a setting. Additionally, I noticed that with `kmenuconfig` you would not notice the appearance/disappearance of bool option depending if COM ports a toggled on/off.

skuenzer commented on drivers/ukconsole/ns16550/Config.uk:
```diff
@@ -1,16 +1,141 @@
+	bool "Initialize early console"
+	depends on (LIBUKCONSOLE_NS16550_COM1 || LIBUKCONSOLE_NS16550_COM2 || LIBUKCONSOLE_NS16550_COM3 || LIBUKCONSOLE_NS16550_COM4)
+
+choice
```
We need an if here to hide the menu if Early console is not actiavted:
```
if LIBUKCONSOLE_NS16550_COM_EARLY_CONSOLE
```

skuenzer commented on drivers/Config.uk:
```diff
@@ -1,3 +1,9 @@
+menu "Serial console"
+
+source "$(shell,$(UK_BASE)/support/build/config-submenu.sh -q -o '$(KCONFIG_DIR)/drivers-console.uk' -r '$(KCONFIG_DRIV_BASE)/ukconsole' -l '$(KCONFIG_DRIV_BASE)/ukconsole' -e '$(KCONFIG_EXCLUDEDIRS)')"
```
Note that `drivers/ukconsole/Config.uk` must not be provided, otherwise the menu includes things twice.

skuenzer commented on plat/kvm/Config.uk:
```diff
@@ -109,8 +112,10 @@ config KVM_VMM_FIRECRACKER
+	imply LIBUKCONSOLE_NS16550 if (ARCH_ARM_64 || ARCH_X86_64)
 	imply LIBUKCONSOLE_NS16550_EARLY_CONSOLE if ARCH_ARM_64
+	imply LIBUKCONSOLE_NS16550_COM1 if ARCH_X86_64
+	imply LIBUKCONSOLE_NS16550_COM_EARLY_CONSOLE if ARCH_X86_64
```
Same here, but `LIBUKCONSOLE_NS16550_COM_EARLY_CONSOLE` needs to be renamed to `LIBUKCONSOLE_NS16550_EARLY_CONSOLE`:
```
	imply LIBUKCONSOLE_NS16550 if (ARCH_ARM_64 || ARCH_X86_64)
	imply LIBUKCONSOLE_NS16550_EARLY_CONSOLE
	imply LIBUKCONSOLE_NS16550_COM1 if ARCH_X86_64
```

skuenzer commented on plat/kvm/Config.uk:
```diff
@@ -109,8 +112,10 @@ config KVM_VMM_FIRECRACKER
 	select HAVE_RANDOM
 	select VIRTIO_MMIO_LINUX_COMPAT_CMDLINE
-	imply LIBUKCONSOLE_NS16550 if ARCH_ARM_64
+	imply LIBUKCONSOLE_NS16550 if (ARCH_ARM_64 || ARCH_X86_64)
```
Due to the `depends on` of `KVM_VMM_FIRECRACKER`, `ARCH_ARM_64` and `ARCH_X86_64` should be already covered here.

skuenzer commented on plat/kvm/Config.uk:
```diff
@@ -97,6 +97,9 @@ config KVM_VMM_QEMU
 	imply LIBUKCONSOLE_PL011_EARLY_CONSOLE if ARCH_ARM_64
+	imply LIBUKCONSOLE_NS16550 if ARCH_X86_64
+	imply LIBUKCONSOLE_NS16550_COM1 if ARCH_X86_64
+	imply LIBUKCONSOLE_NS16550_COM_EARLY_CONSOLE if ARCH_X86_64
```
When seeing this, I am thinking that we should use the same config name for the early console boolean for X86 and Arm: `LIBUKCONSOLE_NS16550_COM_EARLY_CONSOLE`:
```
	imply LIBUKCONSOLE_NS16550 if (ARCH_ARM_64 || ARCH_X86_64)
	imply LIBUKCONSOLE_NS16550_EARLY_CONSOLE
	imply LIBUKCONSOLE_NS16550_COM1 if ARCH_X86_64
```

michpappas commented on drivers/Config.uk:
```diff
@@ -1,3 +1,9 @@
+menu "Serial console"
```
Just a nit, the rest of the entries are in alphabetical order

michpappas commented:
Also if possible could you please rebase to `staging` for the purpose of testing.

michpappas commented:
Thanks!

Reviewed-by: Michalis Pappas <michalis@unikraft.io>

skuenzer commented on drivers/ukconsole/ns16550/ns16550.c:
```diff
@@ -253,143 +337,140 @@ static int early_init(struct ukplat_bootinfo *bi)
+	console_dev->size = dev->size;
+
+	uk_console_register(&console_dev->dev);
+	uk_pr_info("console: ns16550 (%p)\n", &console_dev->base);
```
Maybe we should add also here the interface number since we should got one after register.
```
uk_pr_info("console %"__PRIu16": ns16550 (%p)\n", console_dev->id, &console_dev->base);
```

skuenzer replied:
👍🏻 as we discussed offline, we adopted to `con%" __PRIu16 ": ..."`

skuenzer reviewed (approved):
Thanks a lot for this contribution! This finally improves the handling of consoles with Unikraft.

Reviewed-by: Simon Kuenzer <simon@unikraft.io>

michpappas commented:
Thanks!

Reviewed-by: Michalis Pappas <michalis@unikraft.io>

michpappas reviewed (approved):
Thanks!

Reviewed-by: Michalis Pappas <michalis@unikraft.io>

skuenzer reviewed (approved):
Approved-by: Simon Kuenzer <simon@unikraft.io>
